### PR TITLE
widget_api: initial skeleton

### PR DESCRIPTION
**Pull request by daniel-abramov**
Hey @jplatte,

today, we had a sync-up with @toger5 on the current state of the Client Widget API implementation in the Rust SDK (we've been working on it for a few weeks). If my understanding is correct (@toger5, please correct me if I'm wrong), you wanted to see some "top-level boilerplate" that you could use in order to develop the FFI part of the Widget Client API inside the Rust SDK and check if it matches your expectations from the interface of the API.

We've implemented the part of the spec that should be sufficient for the integration of Element Call in Element X and currently we're working on the final steps to add proper error handling, ensuring that we did not miss anything, and testing it (@toger5 is on it). The latest state could be seen here: https://github.com/toger5/matrix-rust-sdk/blob/client-widget-api/crates/matrix-sdk/src/widget_api/mod.rs

This PR is a small subset of our branch which only includes the minimalistic "upper layer" of the API. We decided to create this small PR as if our understanding is right, you would prefer to see a simple interface before digging into 1.3k+ LOC PR for now 🙂  Additionally, our implementation in a branch misses some documentation and probably would need a round of `clippy` fixes to make sure that we have a clean implementation that would pass the CI.

### CODE REVIEW

File: crates/matrix-sdk/src/lib.rs [hunk 1]
```diff
@@ -48,6 +48,9 @@ pub mod sync;
 #[cfg(feature = "experimental-sliding-sync")]
 pub mod sliding_sync;
 
+#[cfg(feature = "widget-api")]
+pub mod widget_api;
+
 #[cfg(feature = "e2e-encryption")]
 pub mod encryption;
```
Comment (jplatte) [inline crates/matrix-sdk/src/lib.rs:55]: Can you remove the blank lines between all of the module declarations, and instead have just one to separate the `mod`s from the `pub use`s?

File: crates/matrix-sdk/Cargo.toml [hunk 2]
```diff
@@ -58,6 +58,8 @@ experimental-sliding-sync = [
 
 docsrs = ["e2e-encryption", "sqlite", "sso-login", "qrcode", "image-proc"]
 
+widget-api = []
```
Comment (jplatte) [inline crates/matrix-sdk/Cargo.toml:61]: Can you rename this to `experimental-widget-api` and move it up to the other experimental feature definitions?

File: crates/matrix-sdk/src/widget_api/mod.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+//! Client widget API implementation.
+
+use crate::room::Room as JoinedRoom;
+
+pub mod permissions;
+pub mod widget;
+
+/// Starts a client widget API state machine for a given `widget` in a given joined `room`.
+/// The function returns once the widget is disconnected or any terminal error occurs.
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/mod.rs:9]:
```suggestion
/// The function returns once the widget is disconnected or any terminal error occurs.
///
/// Not implemented yet, currently always panics.
```

File: crates/matrix-sdk/src/widget_api/permissions.rs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+//! Types and traits related to the permissions that a widget can request from a client.
+
+use async_trait::async_trait;
+
+/// A trait that must be implemented by an entity that performs the validation / authorization
+/// of the widget's permissions. Typically, it's a native client that "hosts" the widget.
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/permissions.rs:6]: What's a "client" here? A Matrix client? And what significance does the "native" bit have?

File: crates/matrix-sdk/src/widget_api/permissions.rs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#[async_trait]
+pub trait PermissionsProvider: Send + Sync + 'static {
+    /// Receives a request for given permissions and returns the actual permissions that
+    /// the user (client) granted to a given widget.
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/permissions.rs:10]:
```suggestion
    /// the clients grants to a given widget (usually by prompting the user).
```

If I'm understanding correctly?

File: crates/matrix-sdk/src/widget_api/permissions.rs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#[derive(Debug)]
+pub struct Permissions {
+    /// Types of the messages that a widget wants to be able to fetch.
+    pub read: Vec<MessageType>,
+    /// Types of the messages that a widget wants to be able to send.
+    pub send: Vec<MessageType>,
+}
+
+/// An alias for an actual type of the messages that is going to be used with a permission request.
+pub type MessageType = String;
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/permissions.rs:24]: This is for a Matrix event type, and widgets might interact with both message events and state events, right? If yes, this type alias doesn't need to exist, we can use `TimelineEventType` from `ruma::events`.

File: crates/matrix-sdk/src/widget_api/widget.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    /// Whether or not the widget should be initialized on load message (`ContentLoad` message),
+    /// or upon creation/attaching of the widget to the SDK's state machine that drives the API.
+    pub init_on_load: bool,
+}
+
+/// Communication "pipes" with a widget.
+#[derive(Debug)]
+pub struct Comm {
+    /// Raw incoming messages from the widget (normally, formatted as JSON).
+    pub from: Receiver<String>,
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/widget.rs:28]:
```suggestion
    /// Raw incoming messages from the widget (normally formatted as JSON).
    pub from: Receiver<String>,
```

So it's not required for the communication to be JSON?

File: crates/matrix-sdk/src/widget_api/widget.rs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    /// Raw outgoing messages from the client (SDK) to the widget (normally, formatted as JSON).
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/widget.rs:29]:
```suggestion
    /// Raw outgoing messages from the client (SDK) to the widget (normally formatted as JSON).
```

File: crates/matrix-sdk/src/widget_api/widget.rs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+/// Information about a widget.
+#[derive(Debug)]
+pub struct Info {
+    /// Widget's unique identifier.
+    pub id: String,
+    /// Whether or not the widget should be initialized on load message (`ContentLoad` message),
+    /// or upon creation/attaching of the widget to the SDK's state machine that drives the API.
+    pub init_on_load: bool,
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/widget.rs:21]: This seems more like a config thing than an "info" thing.. Anyways I'd be willing to merge this module as-is and change it in follow-up PRs, when it's clearer how everything fits together.

File: crates/matrix-sdk/src/widget_api/mod.rs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+//! Client widget API implementation.
+
+use crate::room::Room as JoinedRoom;
+
+pub mod permissions;
+pub mod widget;
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget_api/mod.rs:6]: I think it's nice that the modules are internally kept small, but for the public API it's probably better (more convenient) if the types from these modules are re-exported here and the modules themselves are private.

```suggestion
mod permissions;
mod widget;

pub use self::{
    permissions::{Permissions, PermissionsProvider},
    widget::{Comm, Info, Widget},
};
```

**Review by jplatte — COMMENTED** [review]
Thanks for the PR! I have a lot of feedback, but it should all be pretty straight-forward. If anything is unclear or you disagree with one of my suggestions, please reply instead of changing things in what seems to be the most appropriate way (this may be obvious, but I haven't interacted with you before and have had the latter happen multiple times 😅).

**Comment by daniel-abramov** [comment]
> Thanks for the PR! I have a lot of feedback, but it should all be pretty straight-forward. If anything is unclear or you disagree with one of my suggestions, please reply instead of changing things in what seems to be the most appropriate way (this may be obvious, but I haven't interacted with you before and have had the latter happen multiple times 😅).

Awesome! Btw, based on the feedback so far I very much appreciate the diligence and attention to detail! 👍 I was worried I'm the only one paying attention to such things 😄 

P.S.: Apologies for not running this through `rustfmt` and `clippy`, I was probably too rushed to submit "something".

File: crates/matrix-sdk/src/lib.rs (hunk 1 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/lib.rs:55]: Sure. One question though: should I also remove the empty lines around the `sliding_sync`. Rationale: I added `pub mod widget_api` there since it was following the experimental sliding sync module which also was separated by spaces.

I.e. it was:

```rust
#[cfg(feature = "experimental-oidc")]
pub mod oidc;
pub mod room;
pub mod sync;

#[cfg(feature = "experimental-sliding-sync")]
pub mod sliding_sync;

#[cfg(feature = "e2e-encryption")]
pub mod encryption;
```

So I placed the `pub mod widget_api` right after the `pub mod sliding_sync`.

**Comment by jplatte** [comment]
Oh, maybe you ran rustfmt, but not nightly rustfmt. We use some rustfmt options that are unfortunately still unstable and only applied when running nightly rustfmt. Try `cargo +nightly fmt`.

File: crates/matrix-sdk/Cargo.toml (hunk 2 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/Cargo.toml]: Makes sense! Done.

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 4 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/permissions.rs:6]: Perhaps poor wording on my side. I believe this trait will be implemented by a component that essentially provides the prompt to the user. "Native" is probably confusing here.

Should I reword it to something like: "A trait that must be implemented by a component that provides the functionality of deciding whether a widget is allowed to use certain capabilities (typically by providing a prompt to the user"?

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 5 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/permissions.rs:10]: Yes, sounds right 🙂

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 6 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/permissions.rs:24]: Right (if I remember correctly). I actually wanted to also use stronger types here and had a discussion with @toger5, but as far as I can remember, @toger5's suggestion was that it can be any event type in theory and that there was no simple way to map them (no enum with a function `f: EventType -> String` and `f: String -> EventType` transformation). @toger5, could you chime in? 🙂

File: crates/matrix-sdk/src/widget_api/widget.rs (hunk 9 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/widget.rs:21]: Yeah, indeed.

File: crates/matrix-sdk/src/widget_api/mod.rs (hunk 10 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/mod.rs:6]: Agree 👍

File: crates/matrix-sdk/src/widget_api/widget.rs (hunk 7 above)
Comment (jplatte) [reply crates/matrix-sdk/src/widget_api/widget.rs:28]: Okay, let's keep it as-is for now.

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 4 above)
Comment (jplatte) [reply crates/matrix-sdk/src/widget_api/permissions.rs:6]: Sounds good, but you can leave out "A trait that must be implemented by", and please close the parentheses as well 😉

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 6 above)
Comment (jplatte) [reply crates/matrix-sdk/src/widget_api/permissions.rs:24]: There are other "event" types that aren't part of `TimelineEventType`, but I'm 99.5% certain that those are out of scope for the widget API. They are ephemeral or mutable things like ephemeral room events (typing, receipts), account data and presence. Often when people talk about events, those are implicitly excluded (see also https://github.com/matrix-org/matrix-spec/issues/897).

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 4 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/permissions.rs]: Done!

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 6 above)
Comment (toger5) [reply crates/matrix-sdk/src/widget_api/permissions.rs:24]: Widgets are allowed to introduce new unspecced message types. Just arbitrary strings for their own usecases. Does TimelineEevntType have a "other" enum case. If so this would very much be a better solution. If not it would be a shame if a widget is working with its own timeline event types but the widget driver is giving a parsing error.

Comment (jplatte) [reply crates/matrix-sdk/src/widget_api/permissions.rs:24]: > Does TimelineEevntType have a "other" enum case.

Yes,any string can be converted `.into()` that type. The variant is hidden / opaque, but to check for a custom type, one can also convert the other way using `.to_string()`.

File: crates/matrix-sdk/src/widget_api/mod.rs
Comment (Hywan) [inline crates/matrix-sdk/src/widget_api/mod.rs:1]: I would rename this module `widget` instead of `widget_api`. Just a personal taste compared to the rest of the SDK.

Comment (jplatte) [reply crates/matrix-sdk/src/widget_api/mod.rs:1]: Though in that case, we should not have a `widget` sub-module. Which is fine by me really, the items from that could be part of this module just as well.

File: crates/matrix-sdk/src/widget_api/permissions.rs (hunk 6 above)
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/permissions.rs]: Nice! Updated.

File: crates/matrix-sdk/src/widget_api/mod.rs
Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/mod.rs:1]: I also prefer short and concise names, I think the only reason we chose `widget_api` originally was that we thought that if we call it `widget`, it might be confused with the widget implementation (not the API, in our case the client widget API), but I'm totally fine with renaming it to `widget` if that's your preference folks!

Comment (jplatte) [reply crates/matrix-sdk/src/widget_api/mod.rs:1]: The client widget API *couldn't* live in this crate though, right? In that case, let's do the rename as suggested by Ivan.

Comment (daniel-abramov) [reply crates/matrix-sdk/src/widget_api/mod.rs:1]: Updated.

File: crates/matrix-sdk/src/widget/permissions.rs [hunk 11]
```diff
@@ -0,0 +1,47 @@
+//! Types and traits related to the permissions that a widget can request from a
+//! client.
+
+use async_trait::async_trait;
+
+use crate::ruma::events::{StateEventType, TimelineEventType};
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget/permissions.rs:6]:
```suggestion
use crate::ruma::events::{MessageLikeEventType, StateEventType};
```

File: crates/matrix-sdk/src/widget/permissions.rs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    pub send: Vec<EventFilter>,
+}
+
+/// Different kinds of filters that could be applied to message (whether for
+/// sending or for receiving).
+#[derive(Debug)]
+pub enum EventFilter {
+    /// Filters for the timeline events, the `data` field is used to store the
+    /// `msgtype`.
+    Timeline(FilterContent<TimelineEventType>),
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget/permissions.rs:34]:
```suggestion
    MessageLike(FilterContent<MessageLikeEventType>),
```

File: crates/matrix-sdk/src/widget/permissions.rs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    pub read: Vec<EventFilter>,
+    /// Types of the messages that a widget wants to be able to send.
+    pub send: Vec<EventFilter>,
+}
+
+/// Different kinds of filters that could be applied to message (whether for
+/// sending or for receiving).
+#[derive(Debug)]
+pub enum EventFilter {
+    /// Filters for the timeline events, the `data` field is used to store the
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget/permissions.rs:32]:
```suggestion
    /// Filters for message-like events, the `data` field is used to store the
```

File: crates/matrix-sdk/src/widget/permissions.rs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    /// `msgtype`.
+    Timeline(FilterContent<TimelineEventType>),
+    /// Filters for the state events, the `data` field is used to store the
```
Comment (jplatte) [inline crates/matrix-sdk/src/widget/permissions.rs:35]:
```suggestion
    /// Filters for state events, the `data` field is used to store the
```

**Review by jplatte — COMMENTED** [review]
I think the conclusion from the discussion on Matrix is simply that we should replace timeline events by message-like event in the `EventFilter`?

**Review by jplatte — CHANGES_REQUESTED** [review]
Actually, I am now realizing that this filtering API is also suffering from the confusion between message events and `m.room.message` events. Not every message(-like) event has a `msgtype`, so please can you clarify what the role of the "timeline" filter is? Are widgets only supposed to interact with `m.room.message` events, or state events? (other examples of message-like events would be reactions)

**Comment by daniel-abramov** [comment]
> Actually, I am now realizing that this filtering API is also suffering from the confusion between message events and `m.room.message` events. Not every message(-like) event has a `msgtype`, so please can you clarify what the role of the "timeline" filter is? Are widgets only supposed to interact with `m.room.message` events, or state events? (an example of a different message-like event would be `m.reaction`)

These are very valid questions!

> Are widgets only supposed to interact with m.room.message events, or state events?

From what I understand, it can interact with both state and message-like events, though among message-like events it only handles `m.room.message` event **at the moment** (EC does not need more than that), though it feels like strictly speaking it could interact with any other message-like events as well.

What about this:

```rust
pub struct EventFilter {
    pub event_type: TimelineEventType,
    pub event_kind: EventKind,
}

pub enum EventKind {
    StateEvent {
        key: String,
    },
    MessageLikeEvent {
        msgtype: Option<String>,
    },
}
```

Does this look sensible to you?

**Comment by jplatte** [comment]
I'd propose this instead:


```rust
pub enum EventFilter {
    MessageLike {
        event_type: MessageLikeEventType,
        /// Additional filter for the msgtype, only used for `m.room.message`.
        msgtype: Option<String>,
    }
    State {
        event_type: StateEventType,
        state_key: String,
    }
}
```

And, final question: Should the `state_key` be optional too? For example, could a widget ask for permission to write `m.room.member` for *any state key*?

**Comment by daniel-abramov** [comment]
> I'd propose this instead:
> 
> ```rust
> pub enum EventFilter {
>     MessageLike {
>         event_type: MessageLikeEventType,
>         /// Additional filter for the msgtype, only used for `m.room.message`.
>         msgtype: Option<String>,
>     }
>     State {
>         event_type: StateEventType,
>         state_key: String,
>     }
> }
> ```
> 
> And, final question: Should the `state_key` be optional too? For example, could a widget ask for permission to write `m.room.member` for _any state key_?

The proposal makes sense 👍 

Indeed, it can request permissions to read/send any state events. So, in this case, I suppose that the `state_key` should either be `Option<String>`, where `None` would be interpreted as "no limitations, any state key", unless we want to explicitly create an enum with the following variants `StateKey = StateKey { key } | AnyStateKey`. Wdyt?

**Comment by jplatte** [comment]
I think `Option<String>` is good enough, the docs can explain its meaning in detail.

File: crates/matrix-sdk/src/widget/mod.rs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+/// Information about a widget.
+#[derive(Debug)]
+pub struct Info {
+    /// Widget's unique identifier.
+    pub id: String,
+    /// Whether or not the widget should be initialized on load message
+    /// (`ContentLoad` message), or upon creation/attaching of the widget to
+    /// the SDK's state machine that drives the API.
+    pub init_on_load: bool,
```
Comment (toger5) [inline crates/matrix-sdk/src/widget/mod.rs:28]: I think we should call it `init_on_content_loaded` even with the doc text I can see how one interprets it as:
init_on_load -> the widget initializes when the I frame is attached/created/"loaded".
(But we want this bool to define define, that the widget is initialized once it sends us the content loaded action.
